crates/inverted/src/lib.rs

```diff
@@ -0,0 +1,191 @@
+#![allow(clippy::len_without_is_empty)]
+
+use base::distance::DistanceKind;
+use base::index::{IndexOptions, SearchOptions};
+use base::operator::Borrowed;
+use base::operator::Operator;
+use base::scalar::{ScalarLike, F32};
+use base::search::{Collection, Element, Payload, Source, Vectors};
+use base::vector::{VectorBorrowed, VectorKind};
+use common::dir_ops::sync_dir;
+use common::json::Json;
+use common::mmap_array::MmapArray;
+use common::remap::RemappedCollection;
+use quantization::{operator::OperatorQuantization, Quantization};
+use storage::{OperatorStorage, Storage};
+
+use std::collections::BTreeMap;
+use std::fs::create_dir;
+use std::path::Path;
+
+pub trait OperatorInverted: Operator + OperatorQuantization + OperatorStorage {}
+
+impl<T: Operator + OperatorQuantization + OperatorStorage> OperatorInverted for T {}
+
+#[allow(dead_code)]
+pub struct Inverted<O: OperatorInverted> {
+    storage: O::Storage,
+    quantization: Quantization<O>,
+    payloads: MmapArray<Payload>,
+    indexes: Json<Vec<u32>>,
+    offsets: Json<Vec<u32>>,
+    scores: Json<Vec<F32>>,
+}
+
+impl<O: OperatorInverted> Inverted<O> {
+    pub fn create(path: impl AsRef<Path>, options: IndexOptions, source: &impl Source<O>) -> Self {
+        if options.vector.v != VectorKind::SVecf32 {
+            panic!("inverted index only supports `SVecf32` vectors");
+        }
+        if options.vector.d != DistanceKind::Dot {
+            panic!("inverted index only supports `Dot` distance");
+        }
+        let remapped = RemappedCollection::from_source(source);
+        from_nothing(path, options, &remapped)
+    }
+
+    pub fn open(path: impl AsRef<Path>) -> Self {
+        open(path)
+    }
+
+    pub fn vbase<'a>(
+        &'a self,
+        vector: Borrowed<'a, O>,
+        opts: &'a SearchOptions,
+    ) -> (Vec<Element>, Box<(dyn Iterator<Item = Element> + 'a)>) {
+        const ZERO: F32 = F32(0.0);
+        let mut doc_score = vec![ZERO; self.payloads.len()];
+        for (token, _) in vector.to_index_vec() {
+            let start = self.offsets[token as usize];
+            let end = self.offsets[token as usize + 1];
+            for i in (start as usize)..(end as usize) {
+                doc_score[self.indexes[i] as usize] += self.scores[i];
```

Comment:
Why it's not `doc_score[self.indexes[i] as usize] += self.scores[i] * val`?

Reply:
`val` is not used here since the sparse matrix already stores the BM25 score. Only the indices are used.

Reply:
Do you mean `vector` is a binary vector?

If it's not guaranteed, and when there are two segments (one is indexed and one is not indexed), the index returns incorrect results because bruteforce in growing segments performs `* val`.

Reply:
How about checking if `val` is 1.0 for first? If it is used correctly, the slow path is never accessed. If it is used incorrectly, it still returns correct results.

Reply:
I think `pg_bestmatch` can change the doc score to bm25 and the query score to 0-1. Both dot-production and inverted-index work in the same way.